cpp/src/arrow/python/helpers.cc

```diff
@@ -227,12 +227,16 @@ Status UInt64FromPythonInt(PyObject* obj, uint64_t* out) {
 }
 
 bool PyDecimal_Check(PyObject* obj) {
-  // TODO(phillipc): Is this expensive?
-  OwnedRef Decimal;
-  Status status = ImportDecimalType(&Decimal);
-  DCHECK_OK(status);
-  const int32_t result = PyObject_IsInstance(obj, Decimal.obj());
-  DCHECK_NE(result, -1) << " error during PyObject_IsInstance check";
+  static OwnedRef decimal_type;
```

Comment:
Note that this may not play well with multiple Python (sub)interpreters in the same process.

Reply:
Is that because the reference count will be shared across them?

Reply:
No, because depending on how the C `decimal` module is implemented, there may be a separate Decimal type per interpreter. This is currently not the case, though, so perhaps we shouldn't worry about it.

Reply:
Here is an alternative implementation. Which one do you prefer?
```c++
bool PyDecimal_Check(PyObject* obj) {
  OwnedRef decimal_type;
  PyTypeObject* obj_type = Py_TYPE(obj);
  // Fast fail for most types other than Decimal, for speed
  if (strcmp(obj_type->tp_name, "decimal.Decimal")) {
    return false;
  }
  Status status = ImportDecimalType(&decimal_type);
  DCHECK_OK(status);
  DCHECK(PyType_Check(decimal_type.obj()));
  // PyObject_IsInstance() is slower as it has to check for virtual subclasses
  const int result =
      PyType_IsSubtype(obj_type, reinterpret_cast<PyTypeObject*>(decimal_type.obj()));
  DCHECK_NE(result, -1) << " error during PyType_IsSubtype check";
  return result == 1;
}
```

Reply:
I guess it's a question of which is more stable: the C implementation of the decimal module or the way Python names types?

I personally don't feel strongly about either one. If I was forced to choose, I probably stick with the first implementation. It seems less brittle, because it doesn't depend on Python's naming convention, but that intuition could be misleading. Maybe that's a very stable part of Python.

Reply:
If the first implementation fails for someone, they'll notice and tell us.
If the second implementation fails for someone, it will produce mysterious slowdowns.
So I think the first implementation should win, for now at least.

Reply:
Great.